File: bindings/typescript/src/mgep.ts
// MGEP TypeScript bindings — pure DataView/Uint8Array wire layout.
//
// Zero dependencies; runs in Node, Deno, Bun, and browsers. Mirrors the
// C# / Java bindings feature-for-feature:
//   * wire constants + enums
//   * header accessors
//   * core-block decoders for the 8 most-used message types
//   * encoder for NewOrderSingle
//   * flex string parser for optional text fields
//
// MGEP's #[repr(C)] core blocks map directly onto byte offsets here.
// Everything is little-endian; we use the `true` argument on DataView
// getters to enforce LE.

// ═══════════════════════════════════════════════════════════════
// Wire constants
// ═══════════════════════════════════════════════════════════════

export const MAGIC = 0x474D; // "MG"
export const VERSION = 1;
export const FRAME_HEADER_SIZE = 8;
export const FULL_HEADER_SIZE = 32;
export const CORE_BLOCK_OFFSET = 32;

export const DECIMAL_SCALE = 100_000_000n;
export const DECIMAL_NULL = -(2n ** 63n);

export const SCHEMA = {
  Session: 0x0000,
  Trading: 0x0001,
  MarketData: 0x0002,
  Quotes: 0x0003,
  PostTrade: 0x0004,
  Risk: 0x0005,
} as const;

export const MSG = {
  // Trading
  NewOrder: 0x01,
  ExecReport: 0x05,
  CancelReject: 0x06,
  BusinessReject: 0x11,
  // Market data
  BookSnapshotRequest: 0x30,
  BookSnapshotBegin: 0x31,
  BookSnapshotLevel: 0x32,
  BookSnapshotEnd: 0x33,
  // Session
  Heartbeat: 0x05,
  Terminate: 0x08,
  ClockStatus: 0x0e,
} as const;

export const Side = {
  Buy: 1,
  Sell: 2,
} as const;
export type SideValue = (typeof Side)[keyof typeof Side];

export const OrderType = {
  Market: 1,
  Limit: 2,
  Stop: 3,
  StopLimit: 4,
} as const;
export type OrderTypeValue = (typeof OrderType)[keyof typeof OrderType];

export const TimeInForce = {
  Day: 1,
  GTC: 2,
  IOC: 3,
  FOK: 4,
  GTD: 5,
} as const;

export const ExecType = {
  New: 0,
  PartialFill: 1,
  Fill: 2,
  Canceled: 4,
  Replaced: 5,
  Rejected: 8,
} as const;

// ═══════════════════════════════════════════════════════════════
// Decimal helpers
// ═══════════════════════════════════════════════════════════════

export function encodeDecimal(value: number | null): bigint {
  if (value === null || Number.isNaN(value)) return DECIMAL_NULL;
  // Avoid float drift by going through Math.round.
  return BigInt(Math.round(value * Number(DECIMAL_SCALE)));
}

export function decodeDecimal(raw: bigint): number | null {
  if (raw === DECIMAL_NULL) return null;
  return Number(raw) / Number(DECIMAL_SCALE);
}

// ═══════════════════════════════════════════════════════════════
// Header
// ═══════════════════════════════════════════════════════════════

export interface FullHeader {
  magic: number;
  flags: number;
  version: number;
  messageSize: number;
  schemaId: number;
  messageType: number;
  senderCompId: number;
  sequenceNum: bigint;
  correlationId: bigint;
  valid: boolean;
}

export function decodeHeader(buf: Uint8Array): FullHeader | null {
  if (buf.length < FULL_HEADER_SIZE) return null;
  const dv = new DataView(buf.buffer, buf.byteOffset, FULL_HEADER_SIZE);
  const magic = dv.getUint16(0, true);
  const flags = dv.getUint8(2);
  const version = dv.getUint8(3);
  const messageSize = dv.getUint32(4, true);
  const schemaId = dv.getUint16(8, true);
  const messageType = dv.getUint16(10, true);
  const senderCompId = dv.getUint32(12, true);
  const sequenceNum = dv.getBigUint64(16, true);
  const correlationId = dv.getBigUint64(24, true);
  return {
    magic,
    flags,
    version,
    messageSize,
    schemaId,
    messageType,
    senderCompId,
    sequenceNum,
    correlationId,
    valid: magic === MAGIC,
  };
}

// ═══════════════════════════════════════════════════════════════
// Trading core blocks
// ═══════════════════════════════════════════════════════════════

export interface NewOrderSingle {
  orderId: bigint;
  clientOrderId: bigint;
  instrumentId: number;
  side: number;
  orderType: number;
  timeInForce: number;
  price: bigint;
  quantity: bigint;
  stopPrice: bigint;
}
export const NewOrderSingleSize = 48;

export function decodeNewOrder(buf: Uint8Array): NewOrderSingle {
  const dv = new DataView(buf.buffer, buf.byteOffset + CORE_BLOCK_OFFSET, NewOrderSingleSize);
  return {
    orderId: dv.getBigUint64(0, true),
    clientOrderId: dv.getBigUint64(8, true),
    instrumentId: dv.getUint32(16, true),
    side: dv.getUint8(20),
    orderType: dv.getUint8(21),
    timeInForce: dv.getUint16(22, true),
    price: dv.getBigInt64(24, true),
    quantity: dv.getBigInt64(32, true),
    stopPrice: dv.getBigInt64(40, true),
  };
}

export interface ExecutionReport {
  orderId: bigint;
  clientOrderId: bigint;
  execId: bigint;
  instrumentId: number;
  side: number;
  execType: number;
  orderStatus: number;
  price: bigint;
  quantity: bigint;
  leavesQty: bigint;
  cumQty: bigint;
  lastPx: bigint;
  lastQty: bigint;
  transactTime: bigint;
}
export const ExecutionReportSize = 88;

export function decodeExecReport(buf: Uint8Array): ExecutionReport {
  const dv = new DataView(buf.buffer, buf.byteOffset + CORE_BLOCK_OFFSET, ExecutionReportSize);
  return {
    orderId: dv.getBigUint64(0, true),
    clientOrderId: dv.getBigUint64(8, true),
    execId: dv.getBigUint64(16, true),
    instrumentId: dv.getUint32(24, true),
    side: dv.getUint8(28),
    execType: dv.getUint8(29),
    orderStatus: dv.getUint8(30),
    // byte 31: _pad
    price: dv.getBigInt64(32, true),
    quantity: dv.getBigInt64(40, true),
    leavesQty: dv.getBigInt64(48, true),
    cumQty: dv.getBigInt64(56, true),
    lastPx: dv.getBigInt64(64, true),
    lastQty: dv.getBigInt64(72, true),
    transactTime: dv.getBigUint64(80, true),
  };
}

export interface BusinessReject {
  refSeqNum: number;
  refMsgType: number;
  businessReason: number;
  orderId: bigint;
}
export const BusinessRejectSize = 16;

export function decodeBusinessReject(buf: Uint8Array): BusinessReject {
  const dv = new DataView(buf.buffer, buf.byteOffset + CORE_BLOCK_OFFSET, BusinessRejectSize);
  return {
    refSeqNum: dv.getUint32(0, true),
    refMsgType: dv.getUint8(4),
    businessReason: dv.getUint8(5),
    // bytes 6..8: _pad
    orderId: dv.getBigUint64(8, true),
  };
}

// ═══════════════════════════════════════════════════════════════
// Market-data snapshot
// ═══════════════════════════════════════════════════════════════

export interface BookSnapshotBegin {
  requestId: bigint;
  instrumentId: number;
  lastAppliedSeq: bigint;
  levelCount: number;
  snapshotId: bigint;
}
export const BookSnapshotBeginSize = 40;

export function decodeSnapshotBegin(buf: Uint8Array): BookSnapshotBegin {
  const dv = new DataView(buf.buffer, buf.byteOffset + CORE_BLOCK_OFFSET, BookSnapshotBeginSize);
  return {
    requestId: dv.getBigUint64(0, true),
    instrumentId: dv.getUint32(8, true),
    lastAppliedSeq: dv.getBigUint64(16, true),
    levelCount: dv.getUint32(24, true),
    snapshotId: dv.getBigUint64(32, true),
  };
}

export interface BookSnapshotLevel {
  snapshotId: bigint;
  levelIndex: number;
  side: number;
  price: bigint;
  quantity: bigint;
  orderCount: number;
}
export const BookSnapshotLevelSize = 40;

export function decodeSnapshotLevel(buf: Uint8Array): BookSnapshotLevel {
  const dv = new DataView(buf.buffer, buf.byteOffset + CORE_BLOCK_OFFSET, BookSnapshotLevelSize);
  return {
    snapshotId: dv.getBigUint64(0, true),
    levelIndex: dv.getUint32(8, true),
    side: dv.getUint8(12),
    // bytes 13..16: _pad
    price: dv.getBigInt64(16, true),
    quantity: dv.getBigInt64(24, true),
    orderCount: dv.getUint32(32, true),
    // bytes 36..40: _pad2
  };
}

export interface BookSnapshotEnd {
  snapshotId: bigint;
  finalSeq: bigint;
  checksum: bigint;
  levelCount: number;
}
export const BookSnapshotEndSize = 32;

export function decodeSnapshotEnd(buf: Uint8Array): BookSnapshotEnd {
  const dv = new DataView(buf.buffer, buf.byteOffset + CORE_BLOCK_OFFSET, BookSnapshotEndSize);
  return {
    snapshotId: dv.getBigUint64(0, true),
    finalSeq: dv.getBigUint64(8, true),
    checksum: dv.getBigUint64(16, true),
    levelCount: dv.getUint32(24, true),
    // bytes 28..32: _pad
  };
}

// ═══════════════════════════════════════════════════════════════
// Session: ClockStatus
// ═══════════════════════════════════════════════════════════════

export interface ClockStatus {
  source: number;
  quality: number;
  observedAt: bigint;
  lastSync: bigint;
  estimatedDriftNs: bigint;
  referenceClockId: bigint;
}
export const ClockStatusSize = 40;

export function decodeClockStatus(buf: Uint8Array): ClockStatus {
  const dv = new DataView(buf.buffer, buf.byteOffset + CORE_BLOCK_OFFSET, ClockStatusSize);
  return {
    source: dv.getUint8(0),
    quality: dv.getUint8(1),
    // bytes 2..8: _pad
    observedAt: dv.getBigUint64(8, true),
    lastSync: dv.getBigUint64(16, true),
    estimatedDriftNs: dv.getBigUint64(24, true),
    referenceClockId: dv.getBigUint64(32, true),
  };
}

/** Quality code 1 == RegulatoryGrade in clock_discipline.rs. */
export function isRegulatoryGrade(status: ClockStatus): boolean {
  return status.quality === 1;
}

// ═══════════════════════════════════════════════════════════════
// Encoder
// ═══════════════════════════════════════════════════════════════

export interface EncodeOrderParams {
  orderId: bigint;
  clientOrderId: bigint;
  instrumentId: number;
  side: SideValue;
  orderType: OrderTypeValue;
  price: number | null;
  quantity: number;
  timeInForce?: number;
  stopPrice?: number | null;
  senderCompId?: number;
  sequenceNum?: bigint;
  correlationId?: bigint;
}

export function encodeNewOrder(p: EncodeOrderParams): Uint8Array {
  const total = FULL_HEADER_SIZE + NewOrderSingleSize;
  const buf = new Uint8Array(total);
  const dv = new DataView(buf.buffer);

  // Frame header
  dv.setUint16(0, MAGIC, true);
  dv.setUint8(2, 0); // flags
  dv.setUint8(3, VERSION);
  dv.setUint32(4, total, true);
  // Message header
  dv.setUint16(8, SCHEMA.Trading, true);
  dv.setUint16(10, MSG.NewOrder, true);
  dv.setUint32(12, p.senderCompId ?? 1, true);
  dv.setBigUint64(16, p.sequenceNum ?? 1n, true);
  dv.setBigUint64(24, p.correlationId ?? 0n, true);
  // Core block
  dv.setBigUint64(32, p.orderId, true);
  dv.setBigUint64(40, p.clientOrderId, true);
  dv.setUint32(48, p.instrumentId, true);
  dv.setUint8(52, p.side);
  dv.setUint8(53, p.orderType);
  dv.setUint16(54, p.timeInForce ?? TimeInForce.Day, true);
  dv.setBigInt64(56, encodeDecimal(p.price), true);
  dv.setBigInt64(64, encodeDecimal(p.quantity), true);
  dv.setBigInt64(72, encodeDecimal(p.stopPrice ?? null), true);

  return buf;
}

// ═══════════════════════════════════════════════════════════════
// Flex string parser
// ═══════════════════════════════════════════════════════════════

/**
 * Parse an optional flex-block string by field_id. Returns null if
 * absent or malformed.
 *
 * Layout: [count:u16][(id:u16, offset:u16)*][data area]
 * Data: [type:u8][len:u16][utf8 bytes...]
 */
export function parseFlexString(
  buf: Uint8Array,
  coreSize: number,
  fieldId = 1,
): string | null {
  const flexStart = CORE_BLOCK_OFFSET + coreSize;
  if (buf.length < flexStart + 2) return null;
  const dv = new DataView(buf.buffer, buf.byteOffset);
  let count = dv.getUint16(flexStart, true);
  if (count > 32) count = 32; // mirror Rust MAX_FLEX_FIELDS
  const dataStart = flexStart + 2 + count * 4;
  for (let i = 0; i < count; i++) {
    const p = flexStart + 2 + i * 4;
    if (p + 4 > buf.length) return null;
    const fid = dv.getUint16(p, true);
    const foff = dv.getUint16(p + 2, true);
    if (fid === fieldId) {
      const pos = dataStart + foff;
      if (pos + 3 > buf.length) return null;
      if (buf[pos] !== 0x0b) return null; // FlexType::String
      const slen = dv.getUint16(pos + 1, true);
      const start = pos + 3;
      if (start + slen > buf.length) return null;
      const bytes = buf.subarray(start, start + slen);
      return new TextDecoder("utf-8").decode(bytes);
    }
  }
  return null;
}

File: bindings/csharp/Mgep.cs
// MGEP C# bindings — pure-managed wire layout (no native library required).
//
// Scope mirrors the Python binding:
//   * Wire structs for every core block via [StructLayout(LayoutKind.Sequential)]
//   * Encode helpers that pack headers + core bytes
//   * Decode dispatch keyed on (schema_id, message_type)
//   * Flex string parser for optional text fields
//
// Because MGEP's wire format is zero-copy and all core blocks are
// `#[repr(C)]` in Rust, they map directly onto `[StructLayout]`
// C# structs. No C ABI shim is needed — managed code reads the
// bytes directly via `MemoryMarshal`.
//
// Tested against .NET 8. Should work on Mono / Unity too.

using System;
using System.Buffers.Binary;
using System.Runtime.InteropServices;
using System.Text;

namespace MGEP
{
    // ═══════════════════════════════════════════════════════════
    // Constants
    // ═══════════════════════════════════════════════════════════

    public static class Wire
    {
        public const ushort Magic = 0x474D; // "MG"
        public const byte Version = 1;
        public const int FrameHeaderSize = 8;
        public const int FullHeaderSize = 32;
        public const int CoreBlockOffset = 32;

        public const long DecimalScale = 100_000_000L;
        public const long DecimalNull = long.MinValue;
        public const ulong TimestampNull = ulong.MaxValue;
    }

    public static class SchemaId
    {
        public const ushort Session = 0x0000;
        public const ushort Trading = 0x0001;
        public const ushort MarketData = 0x0002;
        public const ushort Quotes = 0x0003;
        public const ushort PostTrade = 0x0004;
        public const ushort Risk = 0x0005;
    }

    public enum Side : byte
    {
        Buy = 1,
        Sell = 2,
    }

    public enum OrderType : byte
    {
        Market = 1,
        Limit = 2,
        Stop = 3,
        StopLimit = 4,
    }

    public enum TimeInForce : ushort
    {
        Day = 1,
        GTC = 2,
        IOC = 3,
        FOK = 4,
        GTD = 5,
    }

    public enum ExecType : byte
    {
        New = 0,
        PartialFill = 1,
        Fill = 2,
        Canceled = 4,
        Replaced = 5,
        Rejected = 8,
    }

    // ═══════════════════════════════════════════════════════════
    // Headers
    // ═══════════════════════════════════════════════════════════

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct FrameHeader
    {
        public ushort Magic;
        public byte Flags;
        public byte Version;
        public uint MessageSize;

        public bool IsValid() => Magic == Wire.Magic;
    }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct MessageHeader
    {
        public ushort SchemaId;
        public ushort MessageType;
        public uint SenderCompId;
        public ulong SequenceNum;
        public ulong CorrelationId;
    }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct FullHeader
    {
        public FrameHeader Frame;
        public MessageHeader Message;
    }

    // ═══════════════════════════════════════════════════════════
    // Trading messages
    // ═══════════════════════════════════════════════════════════

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct NewOrderSingle
    {
        public const int Size = 48;

        public ulong OrderId;
        public ulong ClientOrderId;
        public uint InstrumentId;
        public byte Side;
        public byte OrderType;
        public ushort TimeInForce;
        public long Price;
        public long Quantity;
        public long StopPrice;

        public double PriceAsDouble() => Price == Wire.DecimalNull ? double.NaN : Price / (double)Wire.DecimalScale;
        public double QuantityAsDouble() => Quantity / (double)Wire.DecimalScale;
    }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct ExecutionReport
    {
        public const int Size = 88;

        public ulong OrderId;
        public ulong ClientOrderId;
        public ulong ExecId;
        public uint InstrumentId;
        public byte Side;
        public byte ExecType;
        public byte OrderStatus;
        public byte _Pad;
        public long Price;
        public long Quantity;
        public long LeavesQty;
        public long CumQty;
        public long LastPx;
        public long LastQty;
        public ulong TransactTime;

        public double LastPxAsDouble() => LastPx == Wire.DecimalNull ? double.NaN : LastPx / (double)Wire.DecimalScale;
    }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct BusinessReject
    {
        public const int Size = 16;

        public uint RefSeqNum;
        public byte RefMsgType;
        public byte BusinessReason;
        public ushort _Pad;
        public ulong OrderId;
    }

    // ═══════════════════════════════════════════════════════════
    // Market data: snapshot recovery
    // ═══════════════════════════════════════════════════════════

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct BookSnapshotRequest
    {
        public const int Size = 16;
        public ulong RequestId;
        public uint InstrumentId;
        public uint MaxLevels;
    }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct BookSnapshotBegin
    {
        public const int Size = 40;
        public ulong RequestId;
        public uint InstrumentId;
        public uint _Pad;
        public ulong LastAppliedSeq;
        public uint LevelCount;
        public uint _Pad2;
        public ulong SnapshotId;
    }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct BookSnapshotLevel
    {
        public const int Size = 40;
        public ulong SnapshotId;
        public uint LevelIndex;
        public byte Side;
        public byte _Pad1;
        public byte _Pad2;
        public byte _Pad3;
        public long Price;
        public long Quantity;
        public uint OrderCount;
        public uint _Pad4;
    }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct BookSnapshotEnd
    {
        public const int Size = 32;
        public ulong SnapshotId;
        public ulong FinalSeq;
        public ulong Checksum;
        public uint LevelCount;
        public uint _Pad;
    }

    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    public struct ClockStatus
    {
        public const int Size = 40;
        public byte Source;
        public byte Quality;
        public byte _Pad0;
        public byte _Pad1;
        public byte _Pad2;
        public byte _Pad3;
        public byte _Pad4;
        public byte _Pad5;
        public ulong ObservedAt;
        public ulong LastSync;
        public ulong EstimatedDriftNs;
        public ulong ReferenceClockId;

        public bool IsRegulatoryGrade() => Quality == 1;
    }

    // ═══════════════════════════════════════════════════════════
    // Encoder
    // ═══════════════════════════════════════════════════════════

    public static class Encoder
    {
        /// <summary>Encode a NewOrderSingle into a freshly allocated byte[].</summary>
        public static byte[] EncodeNewOrder(
            ulong orderId,
            ulong clientOrderId,
            uint instrumentId,
            Side side,
            OrderType orderType,
            double price,
            double quantity,
            TimeInForce tif = TimeInForce.Day,
            double stopPrice = double.NaN,
            uint senderCompId = 1,
            ulong sequenceNum = 1,
            ulong correlationId = 0)
        {
            var core = new NewOrderSingle
            {
                OrderId = orderId,
                ClientOrderId = clientOrderId,
                InstrumentId = instrumentId,
                Side = (byte)side,
                OrderType = (byte)orderType,
                TimeInForce = (ushort)tif,
                Price = EncodeDecimal(price),
                Quantity = EncodeDecimal(quantity),
                StopPrice = EncodeDecimal(stopPrice),
            };

            uint total = (uint)(Wire.FullHeaderSize + NewOrderSingle.Size);
            var buf = new byte[total];

            // Frame header
            BinaryPrimitives.WriteUInt16LittleEndian(buf.AsSpan(0, 2), Wire.Magic);
            buf[2] = 0; // flags
            buf[3] = Wire.Version;
            BinaryPrimitives.WriteUInt32LittleEndian(buf.AsSpan(4, 4), total);

            // Message header
            BinaryPrimitives.WriteUInt16LittleEndian(buf.AsSpan(8, 2), SchemaId.Trading);
            BinaryPrimitives.WriteUInt16LittleEndian(buf.AsSpan(10, 2), 0x01); // NewOrderSingle
            BinaryPrimitives.WriteUInt32LittleEndian(buf.AsSpan(12, 4), senderCompId);
            BinaryPrimitives.WriteUInt64LittleEndian(buf.AsSpan(16, 8), sequenceNum);
            BinaryPrimitives.WriteUInt64LittleEndian(buf.AsSpan(24, 8), correlationId);

            // Core block — blit the struct bytes directly.
            MemoryMarshal.Write(buf.AsSpan(Wire.CoreBlockOffset), in core);

            return buf;
        }

        private static long EncodeDecimal(double value)
        {
            if (double.IsNaN(value)) return Wire.DecimalNull;
            return (long)(value * Wire.DecimalScale);
        }
    }

    // ═══════════════════════════════════════════════════════════
    // Decoder
    // ═══════════════════════════════════════════════════════════

    public static class Decoder
    {
        /// <summary>
        /// Decode the message header. Returns <c>null</c> if the buffer
        /// is too short or the magic bytes are wrong.
        /// </summary>
        public static FullHeader? DecodeHeader(ReadOnlySpan<byte> buf)
        {
            if (buf.Length < Wire.FullHeaderSize) return null;
            var hdr = MemoryMarshal.Read<FullHeader>(buf);
            if (!hdr.Frame.IsValid()) return null;
            return hdr;
        }

        /// <summary>Decode a NewOrderSingle core block.</summary>
        public static NewOrderSingle DecodeNewOrder(ReadOnlySpan<byte> buf)
            => MemoryMarshal.Read<NewOrderSingle>(buf.Slice(Wire.CoreBlockOffset, NewOrderSingle.Size));

        /// <summary>Decode an ExecutionReport core block.</summary>
        public static ExecutionReport DecodeExecReport(ReadOnlySpan<byte> buf)
            => MemoryMarshal.Read<ExecutionReport>(buf.Slice(Wire.CoreBlockOffset, ExecutionReport.Size));

        /// <summary>Decode a BusinessReject core block.</summary>
        public static BusinessReject DecodeBusinessReject(ReadOnlySpan<byte> buf)
            => MemoryMarshal.Read<BusinessReject>(buf.Slice(Wire.CoreBlockOffset, BusinessReject.Size));

        /// <summary>Decode a BookSnapshotBegin core block.</summary>
        public static BookSnapshotBegin DecodeSnapshotBegin(ReadOnlySpan<byte> buf)
            => MemoryMarshal.Read<BookSnapshotBegin>(buf.Slice(Wire.CoreBlockOffset, BookSnapshotBegin.Size));

        /// <summary>Decode a BookSnapshotLevel core block.</summary>
        public static BookSnapshotLevel DecodeSnapshotLevel(ReadOnlySpan<byte> buf)
            => MemoryMarshal.Read<BookSnapshotLevel>(buf.Slice(Wire.CoreBlockOffset, BookSnapshotLevel.Size));

        /// <summary>Decode a BookSnapshotEnd core block.</summary>
        public static BookSnapshotEnd DecodeSnapshotEnd(ReadOnlySpan<byte> buf)
            => MemoryMarshal.Read<BookSnapshotEnd>(buf.Slice(Wire.CoreBlockOffset, BookSnapshotEnd.Size));

        /// <summary>Decode a ClockStatus core block.</summary>
        public static ClockStatus DecodeClockStatus(ReadOnlySpan<byte> buf)
            => MemoryMarshal.Read<ClockStatus>(buf.Slice(Wire.CoreBlockOffset, ClockStatus.Size));

        /// <summary>
        /// Parse an optional flex string by field_id. Returns null when
        /// the field is absent or the buffer is malformed.
        /// </summary>
        public static string? ParseFlexString(ReadOnlySpan<byte> buf, int coreSize, ushort fieldId = 1)
        {
            int flexStart = Wire.CoreBlockOffset + coreSize;
            if (buf.Length < flexStart + 2) return null;
            ushort count = BinaryPrimitives.ReadUInt16LittleEndian(buf.Slice(flexStart, 2));
            if (count > 32) count = 32; // mirror Rust MAX_FLEX_FIELDS
            int entries = flexStart + 2;
            int data = entries + count * 4;
            for (int i = 0; i < count; i++)
            {
                int p = entries + i * 4;
                if (p + 4 > buf.Length) return null;
                ushort fid = BinaryPrimitives.ReadUInt16LittleEndian(buf.Slice(p, 2));
                ushort foff = BinaryPrimitives.ReadUInt16LittleEndian(buf.Slice(p + 2, 2));
                if (fid == fieldId)
                {
                    int pos = data + foff;
                    if (pos + 3 > buf.Length) return null;
                    if (buf[pos] != 0x0B) return null; // FlexType::String = 0x0B
                    ushort slen = BinaryPrimitives.ReadUInt16LittleEndian(buf.Slice(pos + 1, 2));
                    int start = pos + 3;
                    if (start + slen > buf.Length) return null;
                    return Encoding.UTF8.GetString(buf.Slice(start, slen));
                }
            }
            return null;
        }
    }
}
